pkg/operator/controller/externaldns/rbac_test.go

```diff
@@ -0,0 +1,625 @@
+					},
+					{
+						APIGroups: []string{""},
+						Resources: []string{"endpoints", "services", "pods", "nodes"},
+						Verbs:     []string{"get", "list", "watch"},
+					},
+				},
+			},
+		},
+		{
+			name: "Exists and as expected",
+			existingObjects: []runtime.Object{
+				&rbacv1.ClusterRole{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: controller.ExternalDNSBaseName,
+					},
+					Rules: []rbacv1.PolicyRule{
+						{
+							APIGroups: []string{"extensions", "networking.k8s.io"},
+							Resources: []string{"ingresses"},
+							Verbs:     []string{"get", "list", "watch"},
+						},
+						{
+							APIGroups: []string{""},
+							Resources: []string{"endpoints", "services", "pods", "nodes"},
+							Verbs:     []string{"get", "list", "watch"},
+						},
+					},
+				},
+			},
+			expectedExist: true,
+			expectedRole: rbacv1.ClusterRole{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: controller.ExternalDNSBaseName,
+				},
+				Rules: []rbacv1.PolicyRule{
+					{
+						APIGroups: []string{"extensions", "networking.k8s.io"},
+						Resources: []string{"ingresses"},
+						Verbs:     []string{"get", "list", "watch"},
+					},
+					{
+						APIGroups: []string{""},
+						Resources: []string{"endpoints", "services", "pods", "nodes"},
+						Verbs:     []string{"get", "list", "watch"},
+					},
+				},
+			},
+		},
+		{
+			name: "Exists and needs to be updated",
+			existingObjects: []runtime.Object{
+				&rbacv1.ClusterRole{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: controller.ExternalDNSBaseName,
+					},
+					Rules: []rbacv1.PolicyRule{
+						{
+							APIGroups: []string{"extensions", "networking.k8s.io"},
+							Resources: []string{"ingresses"},
+							Verbs:     []string{"get", "list", "watch"},
+						},
+						{
+							APIGroups: []string{""},
+							Resources: []string{"endpoints", "services", "pods"},
+							Verbs:     []string{"get", "list", "watch"},
+						},
+					},
+				},
+			},
+			expectedExist: true,
+			expectedRole: rbacv1.ClusterRole{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: controller.ExternalDNSBaseName,
+				},
+				Rules: []rbacv1.PolicyRule{
+					{
+						APIGroups: []string{"extensions", "networking.k8s.io"},
+						Resources: []string{"ingresses"},
+						Verbs:     []string{"get", "list", "watch"},
+					},
+					{
+						APIGroups: []string{""},
+						Resources: []string{"endpoints", "services", "pods", "nodes"},
+						Verbs:     []string{"get", "list", "watch"},
+					},
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			cl := fake.NewFakeClient(tc.existingObjects...)
+			r := &reconciler{
+				client: cl,
+				log:    zap.New(zap.UseDevMode(true)),
+			}
+			gotExist, gotRole, _ := r.ensureExternalDNSClusterRole(context.TODO())
+			if gotExist != tc.expectedExist {
+				t.Errorf("expected cluster roles's exist to be %t, got %t", tc.expectedExist, gotExist)
+			}
+			if reflect.DeepEqual(*gotRole, tc.expectedRole) {
```

Comment:
I guess `expectedExist` is always going to be true, right?  Still, it would be good form to check for nil before dereferencing `gotRole`.

Reply:
Yes, but since it's a part of testCase it can be set to `false`, so better to protect the dereferencing. Changed.